explainer.md

```diff
@@ -36,193 +97,126 @@ self.addEventListener('launch', event => {
-browsing context *browsingContext* to a URL *resource*, if *resource* is in the
-scope of an active service worker, the user agent **MAY** go through the
-`launch` flow instead of the normal navigation behaviour.
-
-(NOTE: I don't understand the details of how new browsing contexts are created
-when using `window.open` or "Open in new window". I'm just going to assume that
-the *browsingContext* is null for navigations that open a new window or tab, and
-figure out the "correct" details later.)
-
-The purpose of this "**MAY**" is to allow the user agent to determine in what
-situations to allow the target site to interfere with the navigation. It
-**SHOULD** generally be done when navigating from a resource outside the target
-service worker scope, but **SHOULD NOT** be done when navigating within the same
-service worker scope. It **SHOULD NOT** be done when the user has explicitly
-indicated the intended browsing context for the navigation (e.g., "Open in new
-tab" or "Open in new window").
-
-If the user agent opts to use the `launch` flow, it should do the following
-instead of the normal navigation algorithm, within the context of *resource*'s
-active service worker:
-
-1. Create a new `LaunchEvent` *event* with `type` set to `"launch"`, `url` set
-   to *resource* and `clientId` set to the service worker client ID of
-   *browsingContext*. (`clientId` is `null` if this navigation would open a new
-   browsing context.)
-2. Fire *event* at the service worker global scope. The event listener is
-   **triggered by user activation**.
-3. If *event* is **cancelled**, abort these steps (and do not continue with the
-   navigation).
-4. If *event*'s [[*handlePending*]] slot is null, abort these steps and proceed
-   with the normal navigation algorithm.
+Alternatively, rather than specifying a `type` attribute and class hierarchy, different types of event could be fired in different cases, such as `launch-navigation` and `launch-file`.
+
+On the other hand, the debate of whether to fire a SW event or to trigger a URL navigation with parameters has come up recently in the Deep Linking API [proposal](https://docs.google.com/document/d/1WzpCnpc1N7WjDJnFmj90-Z5SALI3cSPtNrYuH1EVufg/edit#). In that document, mgiuca@ suggests:
+
+> Shortcuts/actions should open URLs, not fire a special SW event. This fits the Web’s defining characteristic, that “places” inside an app should be addressable via a URL.
+
+### Responding with a Client vs. calling Client.focus()
+
+`fetch` events provide a response via a `FetchEvent.respondWith` function. In a similar way, `launch` events could be designed to call a `LaunchEvent.launchWith` function with a `Client` which should be focused.
+
+The main benefit to this approach is that it would ensure that developers don't forget to focus a client window. The main issue with this is that it removes the flexibility for doing things besides focusing windows. For example, `launch` events may just want to show a notification. 
```

Comment:
It's also inconsistent with the 'notification-click' event.

Nit: Trailing whitespace

Reply:
Good thought. I'm not totally sure if it's a strong reason to take that approach, but it's definitely something to note. It could be argued that we made the wrong choice with the notificationclick design and that we should do something different.

Reply:
There's something to be said for a consistent approach.

Arguably, calling client.focus is more powerful, as the client would be able to open multiple windows (e.g. in the case of multiple files).